src/rgw/rgw_acl_swift.cc

```
@@ -59,56 +59,133 @@ static bool uid_is_public(string& uid)
          sub.compare(".referrer") == 0;
 }
 
-void RGWAccessControlPolicy_SWIFT::add_grants(RGWRados *store, list<string>& uids, int perm)
+#define EUNKNOWN_DESIGNATOR EINVAL
+#define EBADREFERRER	EINVAL
+#define ENO_SUCH_USER	EINVAL
+
+int RGWAccessControlPolicy_SWIFT::add_grants(RGWRados *store, list<string>& uids, int perm)
 {
   list<string>::iterator iter;
+  int result = 0;
+//  size_t pos;
   for (iter = uids.begin(); iter != uids.end(); ++iter ) {
     ACLGrant grant;
     RGWUserInfo grant_user;
     string& uid = *iter;
     if (uid_is_public(uid)) {
       grant.set_group(ACL_GROUP_ALL_USERS, perm);
       acl.add_grant(&grant);
-    } else  {
+    } else if ((/* pos = */ uid.find(':')) != std::string::npos) {
+//      auto firstpart = uid.substr(0, pos);	/* project or .r: */
+//      auto secondpart = uid.substr(pos + 1);	/* user or... */
+      if (uid[0] != '.') {
+	rgw_user user(uid);
+        std::string empty_string;
+	grant.set_canon(user, empty_string, perm);
+      } else {
+	/* might be a HTTP referrer-based acl.  Not now (ever?) */
+// need to check firstpart to see if it matches .referrer cases.
+//	bool is_negative = false;
+//	if (*secondpart == '-') {
+//	  secondpart = secondpart.substr(1);
+//	  is_negative = true;
+//	}
+//	grant.set_referrer(secondpart, perm);
+        ldout(cct, 10) << "Unknown designator: " << uid << dendl;
+	if (!result) result = -EUNKNOWN_DESIGNATOR;
```

why not return these failures immediately?